### PR TITLE
[SPARK-24681][SQL] Verify nested column names in Hive metastore

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr added code to check if nested column names do not include ',', ':', and ';' because Hive metastore can't handle these characters in nested column names;
ref: https://github.com/apache/hive/blob/release-1.2.1/serde/src/java/org/apache/hadoop/hive/serde2/typeinfo/TypeInfoUtils.java#L239

## How was this patch tested?
Added tests in `HiveDDLSuite`.

### CODE REVIEW

**Comment by maropu** [comment]
@gatorsmile

**Comment by maropu** [comment]
@gatorsmile plz give me comments on this? Thanks.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 1]
```diff
@@ -2005,6 +2005,24 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
     }
   }
 
+  test("SPARK-24681 checks if nested column names do not include ',', ':', and ';'") {
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Move it to HiveDDLSuite?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala [hunk 2]
```diff
@@ -138,17 +138,35 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
   }
 
   /**
-   * Checks the validity of data column names. Hive metastore disallows the table to use comma in
-   * data column names. Partition columns do not have such a restriction. Views do not have such
-   * a restriction.
+   * Checks the validity of data column names. Hive metastore disallows the table to use some
+   * special characters (',', ':', and ';') in data column names. Partition columns do not have
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: > in data column names.
->
> in data column names, including nested column names.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala [hunk 3]
```diff
@@ -138,17 +138,35 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
...
-        if (colName.contains(",")) {
-          throw new AnalysisException("Cannot create a table having a column whose name contains " +
-            s"commas in Hive metastore. Table: $tableName; Column: $colName")
+      val invalidChars = Seq(",", ":", ";")
+      def verifyNestedColumnNames(schema: StructType): Unit = schema.foreach { f =>
+        f.dataType match {
+          case st: StructType => verifyNestedColumnNames(st)
+          case _ if invalidChars.exists(f.name.contains) =>
+            throw new AnalysisException("Cannot create a table having a nested column whose name " +
+              s"contains invalid characters (${invalidChars.map(c => s"'$c'").mkString(", ")}) " +
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: something wrong, right?

Comment (maropu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: oh..

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala [hunk 4]
```diff
@@ -138,17 +138,36 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
...
-        if (colName.contains(",")) {
-          throw new AnalysisException("Cannot create a table having a column whose name contains " +
-            s"commas in Hive metastore. Table: $tableName; Column: $colName")
+      val invalidChars = Seq(",", ":", ";")
+      def verifyNestedColumnNames(schema: StructType): Unit = schema.foreach { f =>
+        f.dataType match {
+          case st: StructType => verifyNestedColumnNames(st)
+          case _ if invalidChars.exists(f.name.contains) =>
+            val errMsg = "Cannot create a table having a nested column whose name contains " +
+              s"invalid characters (${invalidChars.map(c => s"'$c'").mkString(", ")}) " +
```
[the remark this answers is not in the document]
Comment (gatorsmile) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: Normally, in this case, what we do is like:
```Scala
val invalidCharsString = invalidChars.map(c => s"'$c'").mkString(", ")
            val errMsg = "Cannot create a table having a nested column whose name contains " +
              s"invalid characters ($invalidCharsString) in Hive metastore. Table: $tableName; " +
              s"Column: ${f.name}"

```

[the remark this answers is not in the document]
Comment (maropu) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: aha, I'll fix, thanks!

**Comment by maropu** [comment]
It seems like Aveo errors, so I’ll trigger when it fixed.

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

**Pull request closed without merging**